File: day5/p2.py
#
#   Code should work but takes too long
#
import time

lines = open("input.txt").readlines()
for i in range(len(lines)):
    lines[i] = lines[i].strip()
    
def single_map(seed:int, dest_start:int, source_start:int, length:int):
    dest_start = int(dest_start)
    source_start = int(source_start)
    length = int(length)
    if (seed >= source_start) & (seed < (source_start + length)):
        return (dest_start + (seed - source_start), True)
    return (seed, False)

def whole_map(seed:int, map:list):
    res_seed = seed
    for m in map:
        res_seed, stop = single_map(seed, m[0], m[1], m[2])
        if stop:
            break
    return res_seed
    

def main():
    # Get maps
    maps = [[] for k in range(7)]
    line_index = 1
    map_index = 0
    while line_index < len(lines):
        #print(lines[line_index])
        line_index += 2
        while (lines[line_index] != ""):
            maps[map_index].append(lines[line_index].split(" "))
            line_index += 1
            if (line_index >= len(lines)):
                break
        map_index += 1
    
    # Get seeds and run maps
    t1 = time.time()
    min_location = None
    seed_ranges = lines[0].split(":")[1].split(" ")[1:]
    i = 0
    while (i < len(seed_ranges)):
        seedstart = int(seed_ranges[i])
        seeds_length   = int(seed_ranges[i+1])
        
        seedtime = time.time()
        
        for seed in range(seedstart, seedstart + seeds_length + 1):
            for map in maps:
                seed = whole_map(seed, map)
            if min_location == None:
                min_location = seed
            elif seed < min_location:
                min_location = seed
        
        print(f'Seeds {int(i/2+1)} calculated, time: {time.time() - seedtime}')
        
        i += 2
    
    t2 = time.time()
    print(f'min: {min_location}, time: {t2 - t1}')

if __name__ == "__main__":
    main()

File: 2023/day5/p2/p2.cs
using System;
using System.Collections.Generic;
using System.IO;
using System.Linq;
using System.Diagnostics;

/// @author kurkilmo
/// @version 06.12.2023
/// <summary>
/// </summary>
public class p2 {
    /// <summary>
    /// </summary>
    public static string[] lines;

    public static void Main() {
        Stopwatch timer = new();
        
        lines = File.ReadAllLines("input.txt");
        // Get maps
        List<List<List<Int64>>> maps = new();
        int mapIndex = 0;
        int lineIndex = 1;
        while (lineIndex < lines.Length) {
            lineIndex += 2;
            while (lines[lineIndex] != "") {
                string[] map = lines[lineIndex].Split(" ");
                List<Int64> mapList = map.Select(s => Int64.Parse(s)).ToList();
                try {
                    maps[mapIndex].Add(mapList);
                }
                catch {
                    maps.Add(new List<List<Int64>>());
                    maps[mapIndex].Add(mapList);
                }
                lineIndex++;
                if (lineIndex >= lines.Length) break;
            }

            mapIndex++;
        }
        
        // Get seeds and run maps
        Int64 minLocation = Int64.MaxValue;
        string[] seedRangesString = lines[0].Split(":")[1].Split(" ");
        Int64[] seedRanges = new Int64[seedRangesString.Length-1];
        for (Int64 i = 1; i < seedRangesString.Length; i++) {
            string s = seedRangesString[i].Trim();
            
            seedRanges[i-1] = Int64.Parse(s);
        }
        Console.WriteLine("Start going through seeds");
        for (Int64 i = 0; i < seedRanges.Length; i+=2) {
            timer.Start();
            Int64 seedStart = seedRanges[i];
            Int64 seedLength = seedRanges[i + 1];

            for (Int64 seed = seedStart; seed <= seedStart + seedLength; seed++) {
                Int64 mapped = seed;
                foreach (List<List<Int64>> map in maps) {
                    mapped = Map(mapped, map);
                }

                if (mapped < minLocation) minLocation = mapped;
            }
            timer.Stop();
            Console.WriteLine($"Seed range done, time: {timer.ElapsedMilliseconds / 1000.0} seconds");
            timer.Reset();
        }
        
        Console.WriteLine($"Min location: {minLocation}");
    }

    public static Int64 Map(Int64 seed, List<List<Int64>> map) {
        Int64 resSeed = seed;
        foreach (List<Int64> m in map) {
            Int64 destStart = m[0];
            Int64 sourceStart = m[1];
            Int64 length = m[2];
            
            if (seed >= sourceStart && seed < (sourceStart + length)) {
                resSeed = (destStart + (seed - sourceStart));
                break;
            }
        }

        return resSeed;
    }
}